results/pkg/api/server/server.go

```diff
@@ -155,54 +163,186 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
 	}
-	// get all results from database
-	rows, err := s.db.Query("SELECT data FROM results")
-	if err != nil {
-		log.Printf("failed to query on database: %v", err)
-		return nil, status.Errorf(codes.Internal, "failed to query results: %v", err)
+
+	// checks and refines the pageSize
+	pageSize := req.GetPageSize()
+	if pageSize < 0 {
+		return nil, status.Error(codes.InvalidArgument, "PageSize should be greater than 0")
+	} else if pageSize == 0 {
+		pageSize = listResultsDefaultPageSize
+	} else if pageSize > listResultsMaximumPageSize {
+		pageSize = listResultsMaximumPageSize
 	}
+	// retrieve PRIMARY KEY from PageToken
+	var startPK uint64
+	pageToken := req.GetPageToken()
+	if pageToken != "" {
+		var err error
+		startPK, err = s.decodeTokenToPrimaryKey(pageToken)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Invalid PageToken: %v", err))
+		}
+	}
+
+	var prg cel.Program
 	var results []*pb.Result
-	for rows.Next() {
-		var b []byte
-		if err := rows.Scan(&b); err != nil {
-			log.Printf("failed to scan a row in query results: %v", err)
-			return nil, status.Errorf(codes.Internal, "failed to read result data: %v", err)
+	var err error
+	// return all results back to users if empty query is given
+	if req.GetFilter() != "" {
+		// filter from all results
+		prg, err = s.env.Program(ast)
+		if err != nil {
+			log.Printf("program construction error: %s", err)
+			return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
 		}
-		r := &pb.Result{}
-		if err := proto.Unmarshal(b, r); err != nil {
-			log.Printf("unmarshaling error: %v", err)
-			return nil, status.Errorf(codes.Internal, "failed to parse result data: %v", err)
+	}
+	results, lastKey, err := s.getFilteredPaginatedResults(pageSize, startPK, prg)
+	if err != nil {
+		return nil, err
+	}
+
+	if has, nextPK := s.getNextPrimaryKey(lastKey); has {
+		nextPageToken := s.encodeResult(nextPK)
+		if nextPageToken != "" {
+			return &pb.ListResultsResponse{Results: results, NextPageToken: nextPageToken}, nil
 		}
-		results = append(results, r)
 	}
+	return &pb.ListResultsResponse{Results: results}, nil
+}
 
-	// return all results back to users if empty query is given
-	if req.GetFilter() == "" {
-		return &pb.ListResultsResponse{Results: results}, nil
+// Check if the result can be reserved.
+func matchCelFilter(r *pb.Result, prg cel.Program) (bool, error) {
+	if prg == nil {
+		return true, nil
+	}
+	for _, e := range r.Executions {
+		var (
+			taskrun     pb.TaskRun
+			pipelinerun pb.PipelineRun
+		)
+		if t := e.GetTaskRun(); t != nil {
```

Comment:
Could this just be `taskRun := e.GetTaskRun()` and have the program handle nils?

Reply:
We did some digging into this off-thread: CEL doesn't play nice with nil pointers for proto types (unclear if this is intentional behavior or a bug). We should probably document why we're doing this.